src/webserver/x509.c

```diff
@@ -139,48 +209,73 @@ bool generate_certificate(const char* certfile, bool rsa, const char *domain)
 	// only one certificate being issued with a given browser. Any new generated
 	// certificate would be rejected by the browser as it would have the same
 	// serial number as the previous one and uniques is violated.
-	unsigned char serial[16] = { 0 };
-	mbedtls_ctr_drbg_random(&ctr_drbg, serial, sizeof(serial));
-	for(unsigned int i = 0; i < sizeof(serial) - 1; i++)
-		serial[i] = '0' + (serial[i] % 10);
-	serial[sizeof(serial) - 1] = '\0';
+	unsigned char serial1[16] = { 0 }, serial2[16] = { 0 };
+	mbedtls_ctr_drbg_random(&ctr_drbg, serial1, sizeof(serial1));
+	for(unsigned int i = 0; i < sizeof(serial1) - 1; i++)
+		serial1[i] = '0' + (serial1[i] % 10);
+	serial1[sizeof(serial1) - 1] = '\0';
+	mbedtls_ctr_drbg_random(&ctr_drbg, serial2, sizeof(serial2));
+	for(unsigned int i = 0; i < sizeof(serial2) - 1; i++)
+		serial2[i] = '0' + (serial2[i] % 10);
+	serial2[sizeof(serial2) - 1] = '\0';
 
 	// Create validity period
-	// Use YYYYMMDDHHMMSS as required by RFC 5280
+	// Use YYYYMMDDHHMMSS as required by RFC 5280 (UTCTime)
 	const time_t now = time(NULL);
 	struct tm tms = { 0 };
-	struct tm *tm = localtime_r(&now, &tms);
+	struct tm *tm = gmtime_r(&now, &tms);
 	char not_before[16] = { 0 };
 	char not_after[16] = { 0 };
 	strftime(not_before, sizeof(not_before), "%Y%m%d%H%M%S", tm);
 	tm->tm_year += 30; // 30 years from now
 	strftime(not_after, sizeof(not_after), "%Y%m%d%H%M%S", tm);
 
-	// Generate certificate
-	printf("Generating new certificate with serial number %s...\n", serial);
-	mbedtls_x509write_crt_set_version(&crt, MBEDTLS_X509_CRT_VERSION_3);
+	// 1. Create CA certificate
+	printf("Generating new CA with serial number %s...\n", serial1);
+	mbedtls_x509write_crt_set_version(&ca_cert, MBEDTLS_X509_CRT_VERSION_3);
+
+	mbedtls_x509write_crt_set_serial_raw(&ca_cert, serial1, sizeof(serial1)-1);
+	mbedtls_x509write_crt_set_md_alg(&ca_cert, MBEDTLS_MD_SHA256);
+	mbedtls_x509write_crt_set_subject_key(&ca_cert, &ca_key);
+	mbedtls_x509write_crt_set_subject_key_identifier(&ca_cert);
+	mbedtls_x509write_crt_set_issuer_key(&ca_cert, &ca_key);
+	mbedtls_x509write_crt_set_authority_key_identifier(&ca_cert);
+	mbedtls_x509write_crt_set_issuer_name(&ca_cert, "CN=pi.hole,O=Pi-hole,C=DE");
```

Comment:
```suggestion
	mbedtls_x509write_crt_set_issuer_name(&ca_cert, "CN=pi.hole,O=Pi-hole,C=DE");
```
I'm not sure where Pi-hole is located, but it might no be "DE".

(Same applies to the next line and 266)

Reply:
> Should we add a log line announcing the creation of a CA here

No, it's not even trivial to transport the CA name over there. The vast majority of users will likely add a permanent exception instead of adding the CA certificate. When they want it, the documentation will tell them how - without the need to read any logs.

---

> Should pihole-FTL --read-x509-key read also the tls_ca.crt?

I don't think so. It is a generic tool to reads a certificate. This is still what it does. The CA is independent.

---

> I'm not sure where Pi-hole is located, but it might no be "DE".

In the end it doesn't matter. We can add whatever we want.